[WIP] User module: Selinux and permissions fixes.

chriskarel opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Fixes #24862
This PR corrects issues when the user module is called to create a home directory or generate SSH keys with a user that already exists.  Now the home and .ssh directories will be created with the correct SELinux types.  And in the case of the home directory, it will also properly have its permissions set to 0700.

If SELinux bindings aren't installed, or SELinux is not enabled, the new context changes will simply not run.  

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
User module

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

For Dennis.  I love you little buddy.
<!--- Paste verbatim command output below, e.g. before and after your change -->

chriskarel committed:
Selinux and permissions fixes.

Properly sets SELinux types for created SSH and Home directories for users that already exist.  Also properly sets permissions for created home directories for users that already exist.

chriskarel committed:
Indentation fix

chriskarel committed:
Test case for user changes.

Test creating home directories and .ssh keys for a user that already exists.  Verify SELinux type, ownership, and permissions.

chriskarel committed:
Skip CI SELinux checks

Skip SELinux checks when python bindings are not installed, or SELinux is disabled.

chriskarel committed:
Install SELinux bindings as part of test.

chriskarel committed:
Update main.yml

chriskarel committed:
Workaround for MacOSX weirdness

chriskarel committed:
Update main.yml

chriskarel committed:
Mac debugging

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

rqelibari commented:
Thank you for your efforts!

Some thoughts on that:
I think the permission and selinux type values should not be hardcoded though but rather an option of the module. Further it should not be necessary to install the Python SELinux bindings. A better solution would be to use `semanage fcontext` command or the like.

chriskarel commented:
You're absolutely correct on the permissions, as different distros have different default permissions.  (world readable home directories?  OK...)  That's going to need some umask and login.defs parsing to work properly.

As far as a setype option, I'm not so sure of.  That would seem to imply that you could invoke the user's module to change the SEType of a home directory for an existing user.  This would involve checking and comparing types, recursively changing if different, re-labelling the SSH directories(which are different), etc.  Could quickly become a large bag of worms for a pretty small edge case.  I would argue that's better served by the file module.

chriskarel committed:
Selinux and Permission changes

Fixed permissions and SELinux types for users that already exist, but are having home directories or .ssh keys created.

chriskarel committed:
Typo

chriskarel committed:
Python3 octal

chriskarel committed:
Typos

chriskarel committed:
whitespace

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel committed:
Update main.yml

chriskarel committed:
Update user.py

bcoca commented:
weird .. no clue how this breaks under OS X

mscherer commented on lib/ansible/modules/system/user.py:
```diff
@@ -695,15 +704,16 @@ def modify_user(self):
                 skeleton = self.skeleton
             else:
-                skeleton = '/etc/skel'
+                self.skeleton = '/etc/skel'
```
If you want to use self.skeleton everywhere, the skeleton assignation 2 lines up need to be removed/changed somehow.

mscherer commented on lib/ansible/modules/system/user.py:
```diff
@@ -726,6 +736,33 @@ def chown_homedir(self, uid, gid, path):
+        try:
+            f = open('/etc/login.defs', 'r')
+            for line in f:
+                if 'UMASK' in line and '#' not in line.split()[0]:
```
While there isn't much risk, this would match "UMASKFOO". The 2nd part regarding '#' is also not correct, since '#' need to be the first non space characters, I would recommend to use strip() and verify the first char.

mscherer commented on lib/ansible/modules/system/user.py:
```diff
@@ -726,6 +736,33 @@ def chown_homedir(self, uid, gid, path):
+            f = open('/etc/login.defs', 'r')
+            for line in f:
+                if 'UMASK' in line and '#' not in line.split()[0]:
+                    maskvalue=line.split()[1]
```
This would fail if UMASK has no value. like:
"UMASK     ". Not sure if that's a valid configuration file.

mscherer commented on lib/ansible/modules/system/user.py:
```diff
@@ -726,6 +736,33 @@ def chown_homedir(self, uid, gid, path):
+            for line in f:
+                if 'UMASK' in line and '#' not in line.split()[0]:
+                    maskvalue=line.split()[1]
+                    self.newumask = int(maskvalue, 8)
```
This would fail with a umask in hexadecimal (using 0x, as mentioned in the file comment) or using regular base 10. Both are unlikely, but better code defensively.

mscherer commented on lib/ansible/modules/system/user.py:
```diff
@@ -726,6 +736,33 @@ def chown_homedir(self, uid, gid, path):
+                    maskvalue=line.split()[1]
+                    self.newumask = int(maskvalue, 8)
+            f.close()
+        except:
```
The wide exception catching is a bit weird, as it would mask all kind of errors. Could it be restricted more ?

mscherer commented on lib/ansible/modules/system/user.py:
```diff
@@ -726,6 +736,33 @@ def chown_homedir(self, uid, gid, path):
+        try:
+            if self.skeleton is None:
+                f = open('/etc/default/useradd', 'r')
+                if 'SKELE' in line and '#' not in line.split()[0]:
```
Same comment regarding parsing as the other file ('SKELE' would match 'SKELETON', '#' might be in other place, and we do not verify that '=' is present before splitting. And the wide 'try'/'except' clause should IMHO be changed.

mscherer commented on lib/ansible/modules/system/user.py:
```diff
@@ -726,6 +736,33 @@ def chown_homedir(self, uid, gid, path):
+            self.newumask = os.umask(0)
+        if self.skeleton is not None:
+            skeleumask = os.stat(self.skeleton).st_mode
+        else:
```
How can it be 'none', since we have defined it before to a default value ?

mscherer commented on lib/ansible/modules/system/user.py:
```diff
@@ -2225,6 +2261,16 @@ def main():
+                cmd = [module.get_bin_path('chcon', False)]
+                if cmd[0] is not None:
+                    cmd.append('-t')
+                    cmd.append('user_home_dir_t')
```
I am not sure we should hardcode this value here (as said in the discussion)

mscherer commented on test/integration/targets/user/tasks/main.yml:
```diff
@@ -91,15 +96,118 @@
+  with_items:
+      - "{{user_test3.home}}"
+  register: homedircheck
+  when: ansible_distribution == 'Fedora' or ansible_distribution == 'CentOS'
```
What about RHEL ?

pilou- commented on lib/ansible/modules/system/user.py:
```diff
@@ -726,6 +736,33 @@ def chown_homedir(self, uid, gid, path):
+            pass
+        try:
+            if self.skeleton is None:
+                f = open('/etc/default/useradd', 'r')
```
`f` variable isn't used in this 2nd `try` block: it seems a `for` loop is missing.

sivel commented:
> weird .. no clue how this breaks under OS X

From the logs, the task `try to create a second user (Mac)` specifies `"createhome": false,`

As such, it would seem as though on the 2nd run we specify `"createhome": true,` and is then failing for some reason.  I'm hesitant to run it on my Mac directly, but I'd have to guess there is some issue with trying to do something to the home dir before it's created.

pilou- commented on lib/ansible/modules/system/user.py:
```diff
@@ -726,6 +736,33 @@ def chown_homedir(self, uid, gid, path):
 
+    def parse_defaults(self):
+        try:
+            f = open('/etc/login.defs', 'r')
```
[Context manager](https://docs.python.org/2/library/stdtypes.html#file.close) should be used here (same comment apply to the block below): `with open(...) as file:`.

pilou- commented on lib/ansible/modules/system/user.py:
```diff
@@ -632,6 +632,15 @@ def ssh_key_gen(self):
             try:
                 os.mkdir(ssh_dir, int('0700', 8))
                 os.chown(ssh_dir, info[2], info[3])
+                cmd = [self.module.get_bin_path('chcon', False)]
```
Keyword argument `required` should be passed by keyword instead of by position.

pilou- commented on lib/ansible/modules/system/user.py:
```diff
@@ -2225,6 +2261,16 @@ def main():
                 user.create_homedir(user.home)
                 user.chown_homedir(info[2], info[3], user.home)
+                user.chmod_homedir(user.home)
+                cmd = [module.get_bin_path('chcon', False)]
```
The path of `chcon` could be passed to `ssh_key_gen` in order to avoid to call `get_bin_path` twice with same arguments.

pilou- commented on lib/ansible/modules/system/user.py:
```diff
@@ -632,6 +632,15 @@ def ssh_key_gen(self):
                 os.chown(ssh_dir, info[2], info[3])
+                cmd = [self.module.get_bin_path('chcon', False)]
+                if cmd[0] is not None:
+                    cmd.append('-t')
```
Long name option (`--type` instead of `-t`) could be used here.

pilou- commented on lib/ansible/modules/system/user.py:
```diff
@@ -695,15 +704,16 @@ def modify_user(self):
                 skeleton = self.skeleton
             else:
-                skeleton = '/etc/skel'
+                self.skeleton = '/etc/skel'
```
A default value for `skeleton` is used here. This default value must be mentioned in `skeleton` parameter description (in the documentation part, at the beginning of this file). Better, `/etc/skel` should be the default value of the `skeleton` parameter.

pilou- replied:
`skeleton` requires `createhome` but this isn't enforced. `required_if` should be used with `AnsibleModule`.

pilou- commented on lib/ansible/modules/system/user.py:
```diff
@@ -695,15 +704,16 @@ def modify_user(self):
         return self.modify_user_usermod()
 
     def create_homedir(self, path):
+        self.parse_defaults()
```
Having two separate methods `get_umask` and `get_skeleton` would be better.

gundalow commented:
@chriskarel Thank you for your PR.
As part of reviewing the backlog of PRs we are looking at PRs where the source branch (ie the branch in your fork) has been removed. This is generally a good indicator that people have moved on to different things.

Given that:
* The branch no longer exists
* last comments on this were in May 2017
* Branch is marked as Work in Progress

I'm going to close this.
If you, or anyone else wants to continue with this work then please do feel free to create a fresh PR and `@mention` the previous reviewers in here.
